File: 2015/ruby/day3/part2.rb
#!/usr/bin/ruby

begin
    File.open('input.txt', 'r') do |file|
        position1 = [0, 0]
        position2 = [0, 0]
        visited = [[0, 0]]

        file.each_char.with_index do |char, index|
            position = position1
            if index.odd? then
                position = position2
            end

            case char
            when '^'
                position[1] -= 1
            when '>'
                position[0] += 1
            when 'v'
                position[1] += 1
            when '<'
                position[0] -= 1
            end

            if (visited.detect {|x, y| x == position[0] and y == position[1]}).nil? then
                visited.push(position.clone)
            end
        end

        puts visited.count
    end
rescue SystemCallError
    puts 'input.txt not found.'
end

File: 2016/python/day18/part2.py
#!/usr/bin/python

with open('input.txt') as file_handler:
    row = '.' + file_handler.read().rstrip() + '.'

safe = row.count('.') - 2
traps = ['^^.', '.^^', '^..', '..^']

for _ in range(399999):
    next_row = '.'

    for index in range(1, len(row) - 1):
        next_row += '^' if row[index - 1: index + 2] in traps else '.'

    row = next_row + '.'
    safe += row.count('.') - 2

print safe
